src/pages/index.js

```diff
@@ -51,45 +104,155 @@ const addNewCardPopup = new PopupWithForm(
 //Preview Card Instance
 const cardPreviewPopup = new PopupWithImage("#preview-modal");
 
+//Form Validators
 const editFormValidator = new FormValidator(
   validationSettings,
   profileEditForm
 );
 const addFormValidator = new FormValidator(validationSettings, addNewCardForm);
 
+const avatarFormValidator = new FormValidator(
+  validationSettings,
+  avatarEditForm
+);
+
 editFormValidator.enableValidation();
 addFormValidator.enableValidation();
+avatarFormValidator.enableValidation();
+
+//Delete Card Instance
+const deleteCardPopup = new PopupWithConfirmation("#delete-modal");
+
+//Avatar Edit Instance
+const avatarEditPopup = new PopupWithForm(
+  selectors.avatarPopupSelector,
+  handleAvatarFormSubmit
+);
 
 // Functions
+let newCardSection;
+let userId;
+
+function handleAvatarFormSubmit(inputValues) {
+  avatarEditPopup.renderLoading(true);
+  api
+    .updateProfileAvatar(inputValues.avatar)
+    .then(() => {
+      userInfo.setAvatar(inputValues.avatar);
+      avatarEditPopup.close();
+    })
+    .catch((err) => {
+      console.log(err);
+    })
+    .finally(() => {
+      avatarEditPopup.renderLoading(false, "Save");
+    });
+}
 
 function handleProfileFormSubmit(inputValues) {
-  const { title, description } = inputValues;
-  userInfo.setUserInfo(title, description);
-  profileEditPopup.close();
+  profileEditPopup.renderLoading(true);
+  api
+    .updateProfileInfo(inputValues)
+    .then(() => {
+      userInfo.setUserInfo(inputValues);
+      profileEditPopup.close();
+    })
+    .catch((err) => {
+      console.log(err);
+    })
+    .finally(() => {
+      profileEditPopup.renderLoading(false, "Save");
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```